pennylane/utils.py

```diff
@@ -433,3 +434,77 @@ def expand_vector(vector, original_wires, expanded_wires):
     expanded_tensor = np.moveaxis(expanded_tensor, original_indices, wire_indices)
 
     return expanded_tensor.reshape(2 ** M)
+
+
+class Spy:
+    """Context in which calls to a QubitDevice's execute method are counted without performing the execution.
+
+    The class mocks the execute method, returning a dummy value of the correct dimensions.
+
+    **Example**
+
+    .. code-block:: python
+
+        import pennylane as qml
+        from pennylane.utils import Spy
+        qml.enable_tape()
+
+        dev = qml.device("default.qubit", wires=1)
+
+        @qml.qnode(dev)
+        def circuit(w):
+            qml.RX(w, wires=0)
+            qml.Hadamard(wires=0)
+            return qml.expval(qml.PauliZ(wires=0))
+
+        with Spy() as counter:
+            circuit(0.1)
+            circuit(0.5)
+
+        print(counter.counts) # 2
+
+        with Spy() as counter_grad:
+            g = qml.grad(circuit)
+            g(0.1)
+
+        # parameter-shift rule uses 2+1 executions
+        # to compute the gradient
+        print(counter_grad.counts) # 3
+
+    """
+
+    def __init__(self):
+        self.counts = None
+        self.temp_original = None
+
+    def __enter__(self):
+        # remember the original method
+        self.temp_original = qml._qubit_device.QubitDevice.execute
+
+        # Mock qubit device's execute method.
+        # The side effect will define the output of the mock.
+        qml._qubit_device.QubitDevice.execute = Mock(side_effect=self.side_effect)
```

Comment:
My understanding of context managers is still a bit fuzzy. To clarify what's happening, is it the case that on entry, the spy swaps out the regular `execute` for the "fake" `execute` that just counts the executions? And on exit it replaces it?

Reply:
Yes exactly!